server.js

```diff
@@ -15,40 +24,88 @@ app.use("/public", express.static(__dirname + "/public"));
 app.use(cookieParser());
 
 // get requests
-app.get("/", (req, res) => {
+app.get("/", async (req, res) => {
     const { cookies } = req
-    res.render('index', sessions[cookies.session_id]);
+    try {
+        const currentSession = await Session.findById(cookies.session_id)
+        if (currentSession === null) {
+            res.render('index');
+        } else {
+            res.render('index', currentSession);
+        }
+
+    } catch (err) {
+        res.status(500).json({ result: 'bad', error: err.message })
+    }
+
 })
 // post requests
 app.post("/signup", async (req, res) => {
+    const signupUser = await User.findOne({ username: req.body.username })
+    try {
+        if (signupUser) {
+            console.log("user not added");
+            res.redirect('/');
+        } else {
+            const newUser = await User.create({
+                username: req.body.username,
+                password: req.body.password,
+            });
+            console.log("user added");
+            res.redirect('/');
+        }
 
-    users[req.body.username] = req.body.password;
-    res.redirect('/');
+    } catch (err) {
+        res.status(500).json({ result: 'bad', error: err.message })
+    }
+
+    // users[req.body.username] = req.body.password;
+    // res.redirect('/');
 })
 
 app.post("/login", async (req, res) => {
-
-    if (users[(req.body.username)]) {
-        if (users[req.body.username] === req.body.password) {
-            const session_id = Math.floor(Math.random() * 1000000);
-            sessions[session_id] = { user: req.body.username }
-            res.cookie('session_id', session_id);
-            res.redirect('/');
+    const loginuser = await User.findOne({ username: req.body.username })
+    try {
+        if (loginuser) {
+            if (loginuser.password === req.body.password) {
+                const newSession = await Session.create({
+                    user: req.body.username,
+                });
+                console.log(newSession.id);
+                console.log("session added");
+                res.cookie('session_id', newSession.id);
+            } else {
+                res.redirect('/');
+            }
         } else {
-            res.redirect('/');
+            res.render('index', sessions);
```

Comment:
Why different behavior between when username is wrong and when password is wrong? Both should give an error.

Also, what is the `sessions` object? Aren't you getting your sessions from the DB now?

Reply:
both are not working for some reason. The server crushes and gives: ErrorCaptureStackTrace(err);